botocore/utils.py

```diff
@@ -794,5 +799,102 @@ def _switch_hosts(request, new_endpoint, use_new_scheme=True):
+        self._cache = {}
+
+    def register(self, emitter):
+        emitter.register('needs-retry.s3', self.redirect_from_error)
+        emitter.register('before-call.s3', self.set_request_url)
+        emitter.register('before-parameter-build.s3', self.redirect_from_cache)
+
+    def redirect_from_error(self, request_dict, response, context, **kwargs):
+        """
+        An S3 request sent to the wrong region will return an error that
+        contains the endpoint the request should be sent to. This handler
+        will add the redirect information to the signing context and then
+        redirect the request.
+        """
+        error = response[1].get('Error', {})
+        if error.get('Code') != 'PermanentRedirect':
+            return
+
+        bucket = error['Bucket']
+        client_region = context.get('client_region')
+        new_region = self.get_region_from_response(response)
+
+        if new_region is None:
+            logger.debug(
+                "S3 client configured for region %s but the bucket %s is not "
+                "in that region and the proper region could not be "
+                "automatically determined." % (client_region, bucket))
+            return
+
+        logger.debug(
+            "S3 client configured for region %s but the bucket %s is in region"
+            " %s; Please configure the proper region to avoid multiple "
+            "unnecessary redirects and signing attempts." % (
+                client_region, bucket, new_region))
+
+        endpoint = self._endpoint_resolver.resolve('s3', new_region)
+        endpoint = endpoint['endpoint_url']
+
+        if context is None:
+            context = {}
+
+        signing_context = {
+            'region': new_region,
+            'bucket': bucket,
+            'endpoint': endpoint
+        }
+        context['signing'] = signing_context
+
+        self._cache[bucket] = signing_context
+        self.set_request_url(request_dict, context)
+
+        # Return 0 so it doesn't wait to retry
+        return 0
+
+    def get_region_from_response(self, response):
+        """
+        There are multiple potential sources for the new region to redirect to,
+        but they aren't all universally available for use. This will try to
+        find region from response elements, but will fall back to calling
+        HEAD on the bucket if all else fails.
+        """
+        # First try to source the region from the headers.
+        http_response = response[0]
+        service_response = response[1]
+        if 'x-amz-bucket-region' in http_response.headers:
+            return http_response.headers['x-amz-bucket-region']
+
+        # Next, check the error body
+        region = service_response.get('Error', {}).get('Region')
+        if region:
+            return region
+
+        # Finally, HEAD the bucket. No other choice sadly.
+        endpoint = 'https://' + response[1]['Error']['Endpoint']
+        head_response = requests.head(endpoint)
```

Comment:
This is at least how Ruby is doing it. You don't need to be signed in to head a bucket. In fact, adding auth is far more brittle because bad auth will cause it to fail. A raw HEAD will always succeed since bucket names are not secret.